File: advent-of-code/2015/04/02/solution.py
import hashlib
import os

def find_advent_coin(secret_key):
    number = 1
    while True:
        text = secret_key + str(number)
        hash_bytes = hashlib.md5(text.encode()).digest()
        # Check if first 3 bytes are zero (6 hex digits)
        if hash_bytes[0] == 0 and hash_bytes[1] == 0 and hash_bytes[2] == 0:
            return number
        number += 1

# Test cases - skipped for Part 2 (would take too long)

# Read input and calculate result
script_dir = os.path.dirname(os.path.abspath(__file__))
input_path = os.path.join(script_dir, "..", "input.txt")
with open(input_path) as f:
    secret_key = f.read().strip()

result = find_advent_coin(secret_key)
print(f"Python: {result}")

File: advent-of-code/2015/04/02/solution.go
package main

import (
	"crypto/md5"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func findAdventCoin(secretKey string) int {
	number := 1
	for {
		text := fmt.Sprintf("%s%d", secretKey, number)
		hash := md5.Sum([]byte(text))
		// Check if first 3 bytes are zero (6 hex digits)
		if hash[0] == 0 && hash[1] == 0 && hash[2] == 0 {
			return number
		}
		number++
	}
}

func main() {
	// Test cases - skipped for Part 2 (would take too long)

	// Read input and calculate result
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	inputPath := filepath.Join(dir, "..", "input.txt")
	content, err := os.ReadFile(inputPath)
	if err != nil {
		panic(err)
	}
	secretKey := strings.TrimSpace(string(content))

	result := findAdventCoin(secretKey)
	fmt.Printf("Go: %d\n", result)
}
